src/util/strings.rs

```diff
@@ -0,0 +1,321 @@
+        res.extend_from_slice(self.as_bytes());
+        res
+    }
+
+    fn deserialize(buf: &Vec<u8>, index_ptr: &mut u32, max_size: u32) -> Result<ClarityName, net_error> {
+        let mut index = *index_ptr;
+        let bytes = read_clarity_string_bytes(buf, &mut index, max_size)?;
+
+        // must encode a valid string
+        let s = String::from_utf8(bytes.clone())
+            .map_err(|_e| net_error::DeserializeError)?;
+
+        // must decode to a clarity name
+        let name = ClarityName::try_from(s).map_err(|_e| net_error::DeserializeError)?;
+        *index_ptr = index;
+        Ok(name)
+    }
+}
+
+impl StacksMessageCodec for ContractName {
+    fn serialize(&self) -> Vec<u8> {
+        let mut res = vec![];
+        // ContractName can't be longer than vm::representations::MAX_STRING_LEN, which itself is
+        // a u8, so we should be good here.
+        assert!(self.as_bytes().len() <= CLARITY_MAX_STRING_LENGTH as usize);
+        res.push(self.as_bytes().len() as u8);
+        res.extend_from_slice(self.as_bytes());
+        res
+    }
+
+    fn deserialize(buf: &Vec<u8>, index_ptr: &mut u32, max_size: u32) -> Result<ContractName, net_error> {
+        let mut index = *index_ptr;
+        let bytes = read_clarity_string_bytes(buf, &mut index, max_size)?;
+        
+        // must encode a valid string
+        let s = String::from_utf8(bytes.clone())
+            .map_err(|_e| net_error::DeserializeError)?;
+
+        let name = ContractName::try_from(s).map_err(|_e| net_error::DeserializeError)?;
+        *index_ptr = index;
+        Ok(name)
+    }
+}
+
+impl From<ClarityName> for StacksString {
+    fn from(clarity_name: ClarityName) -> StacksString {
+        // .unwrap() is safe since StacksString is less strict
+        StacksString::from_str(&clarity_name).unwrap()
+    }
+}
+
+impl From<ContractName> for StacksString {
+    fn from(contract_name: ContractName) -> StacksString {
+        // .unwrap() is safe since StacksString is less strict
+        StacksString::from_str(&contract_name).unwrap()
+    }
+}
+
+impl StacksString {
+    /// Is the given string a valid Clarity string?
+    pub fn is_valid_string(s: &String) -> bool {
+        s.is_ascii() && StacksString::is_printable(s)
+    }
+
+    pub fn is_printable(s: &String) -> bool {
+        if !s.is_ascii() {
+            return false;
+        }
+        // all characters must be ASCII "printable" characters, excluding "delete".
+        // This is 0x20 through 0x7e, inclusive, as well as '\t' and '\n'
+        // TODO: DRY up with vm::representations
+        for c in s.as_bytes().iter() {
+            if (*c < 0x20 && *c != ('\t' as u8) && *c != ('\n' as u8)) || (*c > 0x7e) {
+                return false;
+            }
+        }
+        true
+    }
+
+    pub fn try_as_clarity_literal(&self) -> Option<Value> {
+        // must parse to a single Clarity literal
+        match lex(&self.to_string()) {
```

Comment:
Clarity values should be serialized / deserialized as JSON, rather than lexed as Clarity code -- if you use `Value::try_deserialize()`, I think that should do the trick.

https://github.com/blockstack/blockstack-core/blob/develop/src/vm/types/serialization.rs#L346

Reply:
This method get used in the `TransactionPayload::ContractCall` variant to parse a `contract-call?`'s arguments, which must be literals (it's also used to represent NFTs in the token-transfer transaction payload, but that is slated to be dropped).  This means whatever serialization we use here will be set in stone once the system goes live.  Do we intend to keep using JSON, or is a more economical wire format planned?

Reply:
My understanding was that we had decided to use JSON. I am certainly open to using a different serialization format --- either way, Values should be serialized and deserialized using that Value serialization function.

Reply:
Ah.  My understanding was that we were using JSON as a stop-gap, but we would use something more economical (and that offers an unambiguous serialization) for the release.  I'll switch to using the `serialize` and `deserialize` routines here either way.

Reply:
Sounds good -- we can revisit it, but as it stands in SIP-004, Clarity values are serialized as JSON, and I think there's probably good reason to revisit it due to bandwidth concerns.

Reply:
For now, I just ended up implementing `StacksMessageCodec` for `Value`, which internally uses `ClaritySerializable::serialize` for turning a `Value` into bytes and back.  The `TransactionPostCondition::Nonfungible` and `TransactionPayload::ContractCall` variants have been updated to use `Value` instead of `StacksString` for representing actual Clarity values, thereby obviating this method altogether.  The `is_clarity_variable()` method in this struct is still needed for determining that the function name in `TransactionPayload::ContractCall` is still well-formed, however.

Reply:
Fixed in 4f34dedd2c8b1ee4475f79c5fd9405e66cf63151.